src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java

```diff
@@ -130,79 +136,132 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
     public static JavaRDD<GATKRead> coordinateSortReads(final JavaRDD<GATKRead> reads, final SAMFileHeader header, final int numReducers) {
         Utils.validate(header.getSortOrder().equals(SAMFileHeader.SortOrder.coordinate), "Header must specify coordinate sort order, but was" + header.getSortOrder());
 
-        // Turn into key-value pairs so we can sort (by key). Values are null so there is no overhead in the amount
-        // of data going through the shuffle.
-        final JavaPairRDD<GATKRead, Void> rddReadPairs = reads.mapToPair(read -> new Tuple2<>(read, (Void) null));
-
-        // do a total sort so that all the reads in partition i are less than those in partition i+1
-        final Comparator<GATKRead> comparator = new ReadCoordinateComparator(header);
-        final JavaPairRDD<GATKRead, Void> readVoidPairs;
-        final JavaRDD<GATKRead> output;
-        if (numReducers > 0) {
-            readVoidPairs = rddReadPairs.sortByKey(comparator, true, numReducers);
-            output = ReadsSparkSource.putPairsInSamePartition(header, readVoidPairs.keys(), new JavaSparkContext(readVoidPairs.context()));
-        } else {
-            readVoidPairs = rddReadPairs.sortByKey(comparator);
-            output = readVoidPairs.keys();
-        }
-        return output;
+        return sort(reads, new ReadCoordinateComparator(header), numReducers);
     }
 
     /**
      * Sorts the given reads in queryname sort order.
+     * This guarantees that all reads that have the same read name are placed into the same partition
      * @param reads the reads to sort
+     * @param header header of the bam, the header is required to have been set to queryname order
      * @param numReducers the number of reducers to use; a value of 0 means use the default number of reducers
      * @return a sorted RDD of reads
      */
-    public static JavaRDD<GATKRead> querynameSortReads(final JavaRDD<GATKRead> reads, final int numReducers) {
-        // Turn into key-value pairs so we can sort (by key). Values are null so there is no overhead in the amount
-        // of data going through the shuffle.
-        final JavaPairRDD<GATKRead, Void> rddReadPairs = reads.mapToPair(read -> new Tuple2<>(read, (Void) null));
+    public static JavaRDD<GATKRead> querynameSortReads(final JavaRDD<GATKRead> reads, SAMFileHeader header, final int numReducers) {
```

Comment:
Hmmm.... I'm not sure what to say about this class... On the one hand you have `sortSamRecordsToMatchHeader()` which sorts based on the header you provide it, but then we have `querynameSortReads()` and `coordinateSortReads()` which exist as seperate methods if you have GATKreads. Perhaps we should make `sortGATKReadsToMatchHeader()` and use a similar switch statement to denote what we haven't implemented and have that method farm out to `querynameSortReads()` or `coordinateSortReads()` to be more consistent? At the very least `sortGATKReadsToMatchHeader()` should exist.

Reply:
removed the individual sorts and replaced with sortReadsAccordingToHeader